File: executor/packages/aithericon-sdk/src/aithericon/_progress.py
"""Progress and phase tracking via IPC."""

from aithericon._client import get_stub

_PHASE_STATUS_MAP = None


def _get_phase_status_map():
    global _PHASE_STATUS_MAP
    if _PHASE_STATUS_MAP is None:
        from aithericon._proto import executor_sidecar_pb2

        _PHASE_STATUS_MAP = {
            "pending": executor_sidecar_pb2.PHASE_STATUS_PENDING,
            "running": executor_sidecar_pb2.PHASE_STATUS_RUNNING,
            "completed": executor_sidecar_pb2.PHASE_STATUS_COMPLETED,
            "failed": executor_sidecar_pb2.PHASE_STATUS_FAILED,
            "skipped": executor_sidecar_pb2.PHASE_STATUS_SKIPPED,
        }
    return _PHASE_STATUS_MAP


def update_progress(fraction, message=None, current_step=0, total_steps=0):
    """Report execution progress.

    Args:
        fraction: Progress fraction between 0.0 and 1.0.
        message: Optional human-readable progress message.
        current_step: Current step number.
        total_steps: Total number of steps.
    """
    stub = get_stub()
    if stub:
        from aithericon._proto import executor_sidecar_pb2

        stub.UpdateProgress(
            executor_sidecar_pb2.UpdateProgressRequest(
                fraction=fraction,
                message=message or "",
                current_step=current_step,
                total_steps=total_steps,
            )
        )


def define_phases(phase_names):
    """Define the execution phases upfront.

    Args:
        phase_names: List of phase name strings.
    """
    stub = get_stub()
    if stub:
        from aithericon._proto import executor_sidecar_pb2

        stub.DefinePhases(
            executor_sidecar_pb2.DefinePhasesRequest(phase_names=phase_names)
        )


def update_phase(phase_name, status, message=None):
    """Update the status of a named phase.

    Args:
        phase_name: Name of the phase (must match one from define_phases).
        status: One of "pending", "running", "completed", "failed", "skipped".
        message: Optional status message.
    """
    stub = get_stub()
    if stub:
        from aithericon._proto import executor_sidecar_pb2

        status_map = _get_phase_status_map()
        stub.UpdatePhase(
            executor_sidecar_pb2.UpdatePhaseRequest(
                phase_name=phase_name,
                status=status_map.get(status, executor_sidecar_pb2.PHASE_STATUS_PENDING),
                message=message or "",
            )
        )

File: executor/crates/executor-domain/src/progress.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

/// Status of an execution phase.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[cfg_attr(feature = "schema", derive(schemars::JsonSchema, utoipa::ToSchema))]
#[serde(rename_all = "snake_case")]
pub enum PhaseStatus {
    Pending,
    Running,
    Completed,
    Failed,
    Skipped,
}

/// A named phase within an execution.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[cfg_attr(feature = "schema", derive(schemars::JsonSchema, utoipa::ToSchema))]
pub struct Phase {
    /// Phase name (unique within an execution).
    pub name: String,

    /// Current status of this phase.
    pub status: PhaseStatus,

    /// Optional message describing what this phase is doing.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub message: Option<String>,

    /// When this phase started.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub started_at: Option<DateTime<Utc>>,

    /// When this phase ended.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub ended_at: Option<DateTime<Utc>>,
}

/// Progress information for an execution.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[cfg_attr(feature = "schema", derive(schemars::JsonSchema, utoipa::ToSchema))]
pub struct Progress {
    /// Overall fraction complete (0.0 to 1.0).
    pub fraction: f64,

    /// Human-readable progress message.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub message: Option<String>,

    /// Current step number.
    #[serde(default)]
    pub current_step: u64,

    /// Total number of steps (0 if unknown).
    #[serde(default)]
    pub total_steps: u64,

    /// Phases within this execution.
    #[serde(default)]
    pub phases: Vec<Phase>,

    /// When progress was last updated.
    pub updated_at: DateTime<Utc>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn progress_serde_roundtrip() {
        let progress = Progress {
            fraction: 0.5,
            message: Some("Training epoch 5/10".into()),
            current_step: 5,
            total_steps: 10,
            phases: vec![Phase {
                name: "training".into(),
                status: PhaseStatus::Running,
                message: Some("Epoch 5".into()),
                started_at: Some(Utc::now()),
                ended_at: None,
            }],
            updated_at: Utc::now(),
        };

        let json = serde_json::to_string(&progress).unwrap();
        let deserialized: Progress = serde_json::from_str(&json).unwrap();
        assert!((deserialized.fraction - 0.5).abs() < f64::EPSILON);
        assert_eq!(deserialized.current_step, 5);
        assert_eq!(deserialized.phases.len(), 1);
        assert_eq!(deserialized.phases[0].status, PhaseStatus::Running);
    }

    #[test]
    fn phase_status_serde() {
        for status in [
            PhaseStatus::Pending,
            PhaseStatus::Running,
            PhaseStatus::Completed,
            PhaseStatus::Failed,
            PhaseStatus::Skipped,
        ] {
            let json = serde_json::to_value(status).unwrap();
            let deserialized: PhaseStatus = serde_json::from_value(json).unwrap();
            assert_eq!(status, deserialized);
        }
    }
}
